### PR TITLE
Solve #120 Evict cache entry if If-Modified-Since request responded

### PR DESCRIPTION
**Pull request by westonized**
#120

Issue: Evict cache entry if `If-Modified-Since` request responded with `HTTP_OK` code and no `Last-Modified` header

Changes: Add test to expose problem, remove cache when no `Last-Modified` header

### CODE REVIEW

**Comment by dmarkov** [comment]
@westonized Thanks, I will ask someone to do a code review here, thanks for your pull request

**Comment by dmarkov** [comment]
@mkordas it's yours, please review

**Comment by mkordas** [comment]
@westonized I'll do a review today

**Comment by mkordas** [comment]
@westonized first of all, can you put reference to issue and summary of changes to PR description?

File: src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java [hunk 1]
```diff
@@ -173,13 +171,17 @@ private Response validateCacheWithServer(
     }
 
     /**
-     * Add response to cache.
+     * Add response to cache,
+     * Update response in cache or
+     * Evict response from cache.
```
Comment (mkordas) [inline src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java]: @westonized is this one sentence? Due to capital letters it looks like 3.

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 2]
```diff
@@ -65,6 +68,11 @@
     private static final String BODY_UPDATED = "Test body updated";
 
     /**
+     * Test body updated 2.
+     * */
+    private static final String BODY_UPDATED_2 = "Test body updated 2";
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized can it be local variable? We rather try to avoid constants

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 3]
```diff
@@ -130,6 +138,69 @@ public void cachesGetRequest() throws Exception {
     }
 
     /**
+     * Must evict any previous cached entry if a new entry does not have a last
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized second word in Javadoc should be `can`

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 2 above)
Comment (westonized) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @mkordas No as Qulice insists on this, and I have merely copied the style already in the tests.

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 3 above)
Comment (westonized) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @mkordas I don't understand. Where should `can` be in that sentence?

File: src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java (hunk 1 above)
Comment (westonized) [inline src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java]: @mkordas Will change

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 2 above)
Comment (westonized) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @mkordas I will do this, it will now not match the style already in these tests. I will raise todos for those.

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 3 above)
Comment (westonized) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @mkordas seems like that should/could be a qulice rule. I will raise issue.

**Comment by westonized** [comment]
@mkordas I have finished applying all of your comments thanks. Please review again.

**Comment by mkordas** [comment]
@westonized checking again...

File: src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java (hunk 1 above)
Comment (mkordas) [inline src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java]: @westonized looks fine now

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 2 above)
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized right, TODO makes sense

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 3 above)
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized it's already submitted: https://github.com/teamed/qulice/issues/662

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 4]
```diff
@@ -56,11 +59,13 @@
 
     /**
      * Test body.
+     * @todo: #120 inline constant to test(s)
      * */
     private static final String BODY = "Test body";
 
     /**
      * Test body updated.
+     * @todo: #120 inline constant to test(s)
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized can you merge this todo with above?

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 5]
```diff
@@ -56,11 +59,13 @@
 
     /**
      * Test body.
+     * @todo: #120 inline constant to test(s)
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized we need to provide more description here. Imagine you are completely new person in project and you receive just this text as a task

Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized maybe we could link to http://www.yegor256.com/2016/05/03/test-methods-must-share-nothing.html so that new person will understand idea quickly?

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 6]
```diff
@@ -130,6 +135,72 @@ public void cachesGetRequest() throws Exception {
...
+     * We can observe this via the If-Modified-Since headers as when the cache
+     * does not contain an entry, this is not present on the request.
+     * @throws Exception If fails
+     */
+    @Test
+    public void doesNotCacheGetRequestIfTheLastModifiedHeaderIsMissing()
+        throws Exception {
+        final String firstresponse = "Body 1";
+        final String secondresponse = "Body 2";
+        final String thirdreponse = "Body 3";
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized looks like compound name: http://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 7]
```diff
@@ -130,6 +135,72 @@ public void cachesGetRequest() throws Exception {
...
+        final Map<String, String> headers = Collections.singletonMap(
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized looks like http://www.yegor256.com/2015/09/01/redundant-variables-are-evil.html

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 8]
```diff
@@ -220,4 +291,33 @@ public void sendsRequestDirectly() throws Exception {
...
+     * Provides a MkQuery matcher that tests if the request contains the
+     * specified header.
+     * @param header The header to look for
+     * @return A matcher which tests for the supplied header
+     */
+    private static Matcher<MkQuery> queryContainingHeader(final String header) {
+        return new BaseMatcher<MkQuery>() {
+            @Override
+            public boolean matches(final Object object) {
+                final MkQuery query = (MkQuery) object;
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized we should inline this one

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 7 above)
Comment (westonized) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @mkordas again, just copying the original style of the tests, so will raise another todo around them

**Comment by westonized** [comment]
@mkordas I have applied your comments thanks, please review once more

**Comment by mkordas** [comment]
@westonized I'm on it

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 9]
```diff
@@ -56,6 +59,9 @@
 
     /**
      * Test body.
+     * @todo: #120 move constants in this file to their tests because tests
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized the puzzle has incorrect format. It should be `@todo #128:XXmin` where XX is number of minutes for a task. First line should be task title. Next lines are detailed task description

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 10]
```diff
@@ -90,6 +96,8 @@ public void requestWithoutHeaderPassed() throws Exception {
 
     /**
      * LastModifiedCachingWire can cache GET requests.
+     * @todo: #120 inline redundant "headers" variable
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized can we make a puzzle to inline all redundant variables in this test, not only `headers`?

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 11]
```diff
@@ -90,6 +96,8 @@ public void requestWithoutHeaderPassed() throws Exception {
...
+     *  see http://www.yegor256.com/2015/09/01/redundant-variables-are-evil.html
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized generally these puzzles are to trivial, it's 5 min in IntelliJ, not enough for full 30 min task. Can we join them together?

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 9 above)
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized I have an idea, can we add in this puzzle also note to make sure that puzzle format is verified on Travis? Currently it is not... Bulid is green but we wouldn't be able to merge this PR.

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 12]
```diff
@@ -220,4 +293,32 @@ public void sendsRequestDirectly() throws Exception {
...
+    }
+
+    /**
+     * Provides a MkQuery matcher that tests if the request contains the
+     * specified header.
+     * @param header The header to look for
+     * @return A matcher which tests for the supplied header
+     */
+    private static Matcher<MkQuery> queryContainingHeader(final String header) {
+        return new BaseMatcher<MkQuery>() {
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized can we use typesafe matcher instead to avoid nasty caching?

File: src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java [hunk 13]
```diff
@@ -173,13 +171,15 @@ private Response validateCacheWithServer(
...
-     * Add response to cache.
+     * Add, update or evict response in cache.
      * @param req The request to be used as key
-     * @param rsp The response to add
+     * @param rsp The response to add/update
      */
     private void updateCache(final Request req, final Response rsp) {
         if (rsp.headers().containsKey(this.scvh)) {
             this.cache.put(req, rsp);
+        } else {
```
Comment (mkordas) [inline src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java]: @westonized the task says: "Evict cache entry if If-Modified-Since request responded with HTTP_OK code and no Last-Modified header". This would suggest that we should have if with these two conditions somewhere... Where it is?

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 9 above)
Comment (westonized) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @mkordas So I have to estimate the time to raise a puzzle?! Why didn't mention this earlier the first time you commented on the duplicated puzzle? The Travis thing sounds like a completely unrelated issue to this puzzles actual task.

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 14]
```diff
@@ -56,6 +59,9 @@
 
     /**
      * Test body.
+     * @todo: #120 move constants in this file to their tests because tests
+     *  must share nothing
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized my comment was about colon. I haven't spotted it before as I trusted green build. Time is optional indeed, but I'm just not sure what is the default value, so I'd recommend to specify it explicitly.   Puzzle validation task is somehow related. Build was green despite broken puzzle.

File: src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java (hunk 13 above)
Comment (westonized) [inline src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java]: @mkordas I did wrongly assume that there's no way `updateCache` is being called in event of non-OK,but turns out I was wrong. In my defense however the task does _not_ specify what I should do WRT eviction in any other case. i.e. the existing tests pass and the new specified scenario condition passes, other scenarios are untested and unspecified edge cases. I can implement this alongside another test, if you're sure that's the desired behavior?

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 14 above)
Comment (westonized) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @mkordas I'm happy to raise things I notice as I understand what I'm talking about. I don't yet understand puzzle validation on Travis, so unless you give me the exact text you want to be put in this PR, I can't do this, it will just cause another day of delay for this task.

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 11 above)
Comment (westonized) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @mkordas Done

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 12 above)
Comment (westonized) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @mkordas Presume you meant casting. Done

File: src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java (hunk 13 above)
Comment (westonized) [inline src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java]: @mkordas Done anyway with new test.

**Comment by westonized** [comment]
@mkordas Done as much as I can, just can't do the Travis puzzle, as no idea what text to put there that would be enough for any one to be able to understand and do the task.

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 15]
```diff
@@ -130,6 +138,120 @@ public void cachesGetRequest() throws Exception {
...
+                .assertBody(
+                    Matchers.equalTo(third)
+            );
+        } finally {
+            container.stop();
+        }
+    }
+
+    /**
+     * LastModifiedCachingWire can resist cache eviction in the event of a non
```
Comment (westonized) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: Awkward language use because second word must be `can`. I would rather say `LastModifiedCachingWire does not evict cache in the event of a non-OK response`

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 16]
```diff
@@ -56,6 +59,11 @@
 
     /**
      * Test body.
+     * @todo #120:15min Clean tests shared fields and redundant variables
+     *  Move constants in this file to their tests because tests must share
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized very nice, thanks

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 15 above)
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized please remeber to always address your messages: http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html#4-avoid-noise-mdash-address-your-comments

Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized for me English here is fine :) Alternative to `can` would be `cannot`

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 17]
```diff
@@ -130,6 +138,120 @@ public void cachesGetRequest() throws Exception {
...
+        }
+    }
+
+    /**
+     * LastModifiedCachingWire can resist cache eviction in the event of a non
+     * OK response without a last modified header.
+     * @throws Exception If fails
+     */
+    @Test
+    public void doesNotEvictCacheGetRequestOnNonOK()
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized this English is a bit awkward though. What does "does not evict cache get request" mean?

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java [hunk 18]
```diff
@@ -220,4 +342,32 @@ public void sendsRequestDirectly() throws Exception {
...
+    }
+
+    /**
+     * Provides a MkQuery matcher that tests if the request contains the
+     * specified header.
+     * @param header The header to look for
+     * @return A matcher which tests for the supplied header
+     */
+    private static Matcher<MkQuery> queryContainingHeader(final String header) {
+        return new TypeSafeMatcher<MkQuery>() {
```
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized much better with typesafe matcher, thanks

**Comment by mkordas** [comment]
@westonized we can just say: "Please also configure `pdd` and `est` in`.travis.yml` as done e.g. in https://github.com/jcabi/jcabi-xml/blob/master/.travis.yml"

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 14 above)
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized see the text below

File: src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java (hunk 13 above)
Comment (mkordas) [inline src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java]: @westonized thanks, looks good

File: src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java (hunk 12 above)
Comment (mkordas) [inline src/test/java/com/jcabi/http/wire/LastModifiedCachingWireTest.java]: @westonized yeah, casting, sorry

**Comment by westonized** [comment]
@mkordas I have applied your comments thanks. I have also added a todo on my second test as I feel it wasn't really specified behavior in the task and it doesn't feel like correct behavior either. I think the test and the code it drove should stay, but someone needs to decide if it's correct for all response codes.

**Comment by westonized** [comment]
@mkordas In case that wasn't clear, it's ready for you to review again

**Comment by mkordas** [comment]
@westonized right, looking into it

**Comment by mkordas** [comment]
@westinized good job, looks nice!

**Comment by mkordas** [comment]
@rultor merge

**Comment by westonized** [comment]
@mkordas thanks very much for the review!

**Comment by mkordas** [comment]
@westonized you're welcome :)

**Comment by mkordas** [comment]
@yegor256 we need your review here

**Comment by mkordas** [comment]
@yegor256 ping

**Comment by mkordas** [comment]
@yegor256 waiting for your review here

**Comment by mkordas** [comment]
@yegor256 ping again

**Comment by mkordas** [comment]
@yegor256 can you take a look here?

**Comment by westonized** [comment]
@mkordas what is the reason rultor need yegor's approval?It always does?

**Comment by mkordas** [comment]
@westonized because @yegor256 is architect in this project

**Comment by mkordas** [comment]
@westonized in every repo you have file `TEAM.md`. And every PR must be accepted first by reviewer and then by architect

**Comment by mkordas** [comment]
@yegor256 ping

**Comment by mkordas** [comment]
@yegor256 ping again

**Comment by mkordas** [comment]
@yegor256 please review this PR

**Comment by mkordas** [comment]
@yegor256 ping

**Comment by mkordas** [comment]
@yegor256 please take a look here

**Comment by mkordas** [comment]
@yegor256  ping

**Comment by mkordas** [comment]
@yegor256 this one is waiting for 17 days already

**Comment by mkordas** [comment]
@yegor256 ping

**Comment by mkordas** [comment]
@yegor256 ping again

**Comment by mkordas** [comment]
@yegor256 please take a look here

**Comment by mkordas** [comment]
@yegor256 ping

**Comment by mkordas** [comment]
@yegor256 ping again

**Comment by mkordas** [comment]
@yegor256 ping

**Comment by yegor256** [comment]
@rultor merge

**Comment by westonized** [comment]
@yegor256 the build failed, this bug: https://github.com/jcabi/jcabi-http/issues/135 So is this successfully merged? as todo's have not become issues and I've not been paid.

**Comment by mkordas** [comment]
@westonized what is the connection between this PR and #135?

**Comment by westonized** [comment]
@mkordas Sorry, I meant to include this link. I got an email about a failed master build while integrating this PR https://travis-ci.org/jcabi/jcabi-http/builds/137975811

That's the #135 issue. And I wonder if everything is ok with the merge as my todo has not been promoted to an issue, and I have not been paid for issue.

**Comment by mkordas** [comment]
@westonized Travis and Rultor are totally separate. Travis is just informational

**Comment by mkordas** [comment]
@westonized and for the rest you need to wait

**Comment by westonized** [comment]
@mkordas ok thanks, it's my first merge, forgive my impatience, but as you know it's taken a month to get this far. Not typical I hope?

**Comment by mkordas** [comment]
@westonized definitely not, the idea is to do things as soon as possible, but it all depends on reviewer and architect availability

**Comment by dmarkov** [comment]
@mkordas thank you, added **1 hour and 15 mins** to your acc, payment num is `5763efed19c8d0014900082e`, 719 hours and 16 mins spent in total to complete... you're getting extra minutes here (c=60)... +75 added to your rating, current score is: [+5596](http://www.netbout.com/b/36080?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy
